### PR TITLE
Refactor GroovycWrapper into two parts

### LINKED ISSUE #84: Split up GroovycWrapper
One part should implement getting compilation errors and generating an AST, and another part should be parsing symbols, references, etc from a given AST.

### PR DESCRIPTION
**Pull request by natgabb**
- One part is the TreeParser and takes care of parsing symbols.
- The other part is the WorkspaceCompiler which takes care of compiling the code.
- The CompilerWrapperProvider is gone and now the CompilerWrapper is just part of the shared state.
- New packages names: ls.groovy.api, ls.groovy.compilation, ls.groovy.services

Closes #84

### CODE REVIEW

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/compilation/GroovyTreeParser.java [hunk 1]
```diff
@@ -66,163 +53,107 @@
...
-     * @param targetDirectory the directory in which to put generated files
+     * @param unitProvider the compilation unit provider in which to store our current unit
      * @param workspaceRoot the directory to compile
      * @param changedFilesRoot the directory in which to temporarily store incrementally changed files
-     * @return the newly created GroovycWrapper
+     * @return the newly created GroovyTreeParser
      */
-    public static GroovycWrapper of(Path targetDirectory, Path workspaceRoot, Path changedFilesRoot) {
-        checkNotNull(targetDirectory, "targetDirectory must not be null");
+    public static GroovyTreeParser of(CompilationUnitProvider unitProvider, Path workspaceRoot, Path changedFilesRoot) {
```
Comment (natgabb) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/compilation/GroovyTreeParser.java]: `GroovyTreeParser` shouldn't need `changedFilesRoot`. I need to figure out another way to resolve uri in `getWorkspaceUri`

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java [hunk 2]
```diff
@@ -31,13 +32,25 @@
 
 public final class GroovyLanguageServerConfig implements LanguageServerConfig {
```
Comment (jiahuijiang) [inline groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java]: Can we rename this class to `GroovyLanguageServerSharedObjects` or something like that
Same with the interface

**Comment by jiahuijiang** [comment]
looks good in general :)

File: groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java (hunk 2 above)
Comment (natgabb) [reply groovy-language-server/src/main/java/com/palantir/ls/groovy/GroovyLanguageServerConfig.java]: How about `GroovyLanguageServerState` ?

**Comment by natgabb** [comment]
@gracew can you please take a look when you get a chance :)

**Comment by natgabb** [comment]
I still need to refactor the tests into two separate files but I think it might be easier to do in a follow up PR.

**Pull request merged**